# 3D CNN

### Loading Cifar10 dataset and reshaping it and Doing Train Validation Split

In [121]:
import sys

#importing 3 party library
import keras
from keras.models import Sequential
from keras.datasets import cifar10 
from keras import backend as K
from keras import optimizers
from keras.layers import Dense, Flatten
from keras.layers import Conv3D, MaxPooling3D,Activation
from scipy.sparse import coo_matrix
from sklearn.utils import shuffle
import numpy as np


(X, Y), (X_test, Y_test)  = cifar10.load_data()



#reshaping the image .
if K.image_data_format == 'channels_first':
    X = X.reshape(X.shape[0],1,32,32,3)
    X_test  = X_test.reshape(X_test.shape[0],1,32,32,3)
else:
    X = X.reshape(X.shape[0],3,32,32,1)
    X_test  = X_test.reshape(X_test.shape[0],3,32,32,1)



X = X.astype('float32')
X_test = X_test.astype('float32') 

X /= 255.0
X_test /= 255.0
Y = keras.utils.np_utils.to_categorical(Y)
Y_test  = keras.utils.np_utils.to_categorical(Y_test)

X_train = X[0:42000]
X_val = X[42000::]

print(X.shape)
print(X_train.shape, 'train samples')
print(X_val.shape, 'validation samples')
print(X_test.shape, 'test samples')




Y_train = Y[0:42000]
Y_val = Y[42000::]

print(Y_train.shape,Y_val.shape,Y_test.shape, 'test samples')

(50000, 3, 32, 32, 1)
(42000, 3, 32, 32, 1) train samples
(8000, 3, 32, 32, 1) validation samples
(10000, 3, 32, 32, 1) test samples
(42000, 10) (8000, 10) (10000, 10) test samples


## Model : 3DCNN with single Hidden Layer ;
### 1. 32 3D Convulutional Kernel Followed by 2X2X1 maxpooling
### 2. Fully Connected layer with 128 Units

In [0]:
def CNN_3D(batchsize,lr):
  
    Model = Sequential()
    Model.add(Conv3D(32,kernel_size = (3,3,3),activation="relu",input_shape = (3,32,32,1),padding='same'))
    Model.add(MaxPooling3D(pool_size=(2,2,1),padding='same'))


    #fully connected layer
    Model.add(Flatten())      
    Model.add(Dense(128,activation="relu"))
    Model.add(Dense(num_classes  , activation="softmax"))

    

    Model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.SGD(lr=0.05),
              metrics=['accuracy'])


    Model.fit(X_train,Y_train, batch_size=batchsize,epochs=epochs,
              validation_data=(X_val, Y_val))

    score = Model.evaluate(X_test, Y_test, verbose=1)
    print('Test loss:', score[0])
    print('Test accuracy:', score[1])

## Batch Size =128 and Learning rate =0.01

In [123]:
epochs=15
CNN_3D(128,0.01)

Train on 42000 samples, validate on 8000 samples
Epoch 1/15
42000/42000 [==============================] - 9s 211us/step - loss: 1.9545 - acc: 0.2943 - val_loss: 2.1673 - val_acc: 0.2316
Epoch 2/15
42000/42000 [==============================] - 8s 188us/step - loss: 1.6533 - acc: 0.4167 - val_loss: 1.7304 - val_acc: 0.3805
Epoch 3/15
42000/42000 [==============================] - 8s 186us/step - loss: 1.5401 - acc: 0.4605 - val_loss: 1.6290 - val_acc: 0.4208
Epoch 4/15
42000/42000 [==============================] - 8s 191us/step - loss: 1.4653 - acc: 0.4856 - val_loss: 2.1044 - val_acc: 0.3028
Epoch 5/15
42000/42000 [==============================] - 8s 187us/step - loss: 1.4012 - acc: 0.5060 - val_loss: 1.6723 - val_acc: 0.4155
Epoch 6/15
42000/42000 [==============================] - 8s 189us/step - loss: 1.3436 - acc: 0.5279 - val_loss: 1.8436 - val_acc: 0.3738
Epoch 7/15
42000/42000 [==============================] - 8s 188us/step - loss: 1.2941 - acc: 0.5460 - val_loss: 1.8731 - v

## Batch size = 128 and Learning rate =0.05

In [127]:
CNN_3D(128,0.05)

Train on 42000 samples, validate on 8000 samples
Epoch 1/15
42000/42000 [==============================] - 9s 224us/step - loss: 1.9400 - acc: 0.3015 - val_loss: 1.7747 - val_acc: 0.3486
Epoch 2/15
42000/42000 [==============================] - 8s 196us/step - loss: 1.6420 - acc: 0.4233 - val_loss: 1.6495 - val_acc: 0.4156
Epoch 3/15
42000/42000 [==============================] - 8s 190us/step - loss: 1.5281 - acc: 0.4611 - val_loss: 1.7288 - val_acc: 0.3874
Epoch 4/15
42000/42000 [==============================] - 8s 187us/step - loss: 1.4493 - acc: 0.4876 - val_loss: 1.6414 - val_acc: 0.4104
Epoch 5/15
42000/42000 [==============================] - 8s 189us/step - loss: 1.3883 - acc: 0.5103 - val_loss: 1.8140 - val_acc: 0.3867
Epoch 6/15
42000/42000 [==============================] - 8s 187us/step - loss: 1.3316 - acc: 0.5321 - val_loss: 1.5558 - val_acc: 0.4427
Epoch 7/15
42000/42000 [==============================] - 8s 186us/step - loss: 1.2770 - acc: 0.5488 - val_loss: 1.6755 - v

## Batch Size =256 and Learning rate =0.01



In [124]:
CNN_3D(256,0.01)

Train on 42000 samples, validate on 8000 samples
Epoch 1/15
42000/42000 [==============================] - 8s 188us/step - loss: 2.0482 - acc: 0.2608 - val_loss: 2.2851 - val_acc: 0.1668
Epoch 2/15
42000/42000 [==============================] - 7s 165us/step - loss: 1.7637 - acc: 0.3790 - val_loss: 2.0144 - val_acc: 0.2931
Epoch 3/15
42000/42000 [==============================] - 7s 166us/step - loss: 1.6479 - acc: 0.4230 - val_loss: 2.2220 - val_acc: 0.2611
Epoch 4/15
42000/42000 [==============================] - 7s 164us/step - loss: 1.5907 - acc: 0.4417 - val_loss: 2.1135 - val_acc: 0.3216
Epoch 5/15
42000/42000 [==============================] - 7s 165us/step - loss: 1.5252 - acc: 0.4657 - val_loss: 1.8044 - val_acc: 0.3588
Epoch 6/15
42000/42000 [==============================] - 7s 165us/step - loss: 1.4798 - acc: 0.4832 - val_loss: 1.9297 - val_acc: 0.3275
Epoch 7/15
42000/42000 [==============================] - 7s 165us/step - loss: 1.4371 - acc: 0.4976 - val_loss: 1.9971 - v

## Batch Size =256 and Learning rate =0.05

In [126]:
CNN_3D(256,0.05)

Train on 42000 samples, validate on 8000 samples
Epoch 1/15
42000/42000 [==============================] - 8s 193us/step - loss: 2.0433 - acc: 0.2642 - val_loss: 2.2590 - val_acc: 0.2018
Epoch 2/15
42000/42000 [==============================] - 7s 165us/step - loss: 1.7500 - acc: 0.3794 - val_loss: 1.9186 - val_acc: 0.3137
Epoch 3/15
42000/42000 [==============================] - 7s 163us/step - loss: 1.6374 - acc: 0.4232 - val_loss: 1.9311 - val_acc: 0.3287
Epoch 4/15
42000/42000 [==============================] - 7s 164us/step - loss: 1.5651 - acc: 0.4482 - val_loss: 1.6596 - val_acc: 0.4149
Epoch 5/15
42000/42000 [==============================] - 7s 164us/step - loss: 1.5057 - acc: 0.4741 - val_loss: 2.0450 - val_acc: 0.3410
Epoch 6/15
42000/42000 [==============================] - 7s 164us/step - loss: 1.4693 - acc: 0.4865 - val_loss: 2.2267 - val_acc: 0.3211
Epoch 7/15
42000/42000 [==============================] - 7s 166us/step - loss: 1.4291 - acc: 0.5001 - val_loss: 1.7072 - v

# Inference

#### For 15 epochs the model didnt converged to good result but among different choices of batch size and Learning rate 
####  Batch size =128 and learning rate =0.01 gave good Test accuracy of 49%

####  With the Batch size of 128 updates are frequent than 256 batch size and The learning rate 0.01 gave stable convergence